### DISCUSSION TITLE
How a program in user mode calls the kernel tests defined in module.

### DISCUSSION THREAD

**Discussion by wenhaopang** [NVIDIA/open-gpu-kernel-modules#403] [Q&A] [answered]
How a program in user mode calls the kernel tests defined in module.
as described on lines 42-49 in `kernel-open/nvidia-uvm/uvm_common.c` 

```
// This parameter allows a program in user mode to call the kernel tests
// defined in this module. This parameter should only be used for testing and
// must not be set to true otherwise since it breaks security when it is
// enabled. By default and for safety reasons this parameter is set to false.
int uvm_enable_builtin_tests __read_mostly = 0;
module_param(uvm_enable_builtin_tests, int, S_IRUGO);
MODULE_PARM_DESC(uvm_enable_builtin_tests,
                 "Enable the UVM built-in tests. (This is a security risk)");
```

**Comment by mtijanic** [comment, marked as answer]
Through `ioctl()`. The [`uvm_ioctl()`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/nvidia-uvm/uvm.c#L698-L750) function lists all the valid production level ioctl codes that can be made, and then it falls through to [`uvm_test_ioctl()`](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/kernel-open/nvidia-uvm/uvm_test.c#L221-L337) which will only run the tests if the module param was specified.

**Reply by wenhaopang** [reply]
Thank you very much, I'm going to learn this part of the code.
